runtime/lua/vim/lsp/util.lua

```diff
@@ -31,6 +31,36 @@ local function split_lines(value)
   return split(value, '\n', true)
 end
 
+function M.add_hook_after(func, new_fn)
+  if func then
+    return function(...)
+      -- TODO which result?
+      func(...)
+      return new_fn(...)
+    end
+  else
+    return new_fn
+  end
+end
+
+function M.wrap_on_init(config)
+  return M.add_hook_after(config.on_init,
```

Comment:
Based on the discussion in the other issue, I wonder if Neovim should just always send the `didChangeConfiguration` notification by default? It sounded as if there isn't really a scenario where a user would want to opt-out of it.


If not and we leave that up to users to opt-in, then I'm not sure if this is the right trade-off between coupling `on_init`/`didChangeConfiguration` and convenience sugar for users.


Currently users would have to add:

```lua
config.on_init = function(client, initialize_result)
  client.notify('workspace/didChangeConfiguration', {settings = client.config.settings})`
  ...
end
```

Instead of:

```lua
config = vim.lsp.util.wrap_on_init(config)
```


I think that the first variant is already simple enough.
But if we wanted to make it even easier and keep the validation aspect we could add something like `vim.lsp.util.make_validated_settings_params()` or `vim.lsp.util.make_configuration_params()` to create the payload. That would follow the existing pattern of having `util` functions to create request params, and it would split up the IO related parts and the pure function parts.


Or remove the `on_init` coupling and create something like this:


```lua
config.on_init = vim.lsp.util.add_did_change_configuration_notify(config.on_init)
```